proposals/p0821.md

```diff
@@ -0,0 +1,1119 @@
+
+The resulting primitives should directly and effectively address each use case,
+not merely replicate the various patterns present in C++.
+
+The only use case that this proposal specifically doesn't address is
+_syntax-free dereference_. Instead, we suggest it shouldn't be a priority for
+Carbon, especially in interface boundaries. Some ideas for steps that could be
+taken in the future here are mentioned at the end.
+
+### Immutable views of values
+
+The default and most primitive construct for both locals and parameters in
+Carbon should be immutable views of values or just _values_ in Carbon. These
+should cover the exact intended use cases of `const` references in C++, but
+should improve on them by enabling in-register parameter passing and other
+relevant optimizations. More generally, they address the _immutable_ use case
+for `const`. So the C++ code:
+
+```
+const int count = 42;
+```
+
+Would become:
+
+```
+let count: i32 = 42;
+```
+
+**Syntax alternatives not currently proposed:**
+
+-   It is appealing to consider the `val` introducer instead of `let` given that
+    these are expected to be called "values". However, there is some concern
+    over the visual similarity of `val` and `var`, and if pronounced as written
+    instead of as "value" ar "variable", the difference relies on a subtle
+    consonant distinction not present in some widely used spoken language. That
+    said, it is worth investigating whether there is a real user preference
+    between these introducers. Kotlin even has both `val` and `var` with a
+    similar distinction and so might serve as the basis for investigating this.
+    For now, using the less controversial spelling is proposed as this is an
+    easy thing to change later.
+-   Another potential introducer would be `const`. However the semantics would
+    be subtly different even if the use cases would be similar, which might make
+    reuse of the keyword more confusing than helpful. Also, it is longer and
+    expected to be extremely common.
+
+#### Function parameters
+
+Most function parameters in C++ are immutable, and passed as either a `const`
+reference to avoid copies or by-value to avoid the indirection of an explicit
+reference. When C++ code passes by-value in these cases users often omit `const`
+even when they could use it, the same way they do for local variables -- C++'s
+requirement to remember an extra qualifier ends up with code failing to document
+where mutability is required. Carbon code is expected to have the same desired
+default, but we want to provide a single, easier default option, especially for
+function parameters. As a consequence, Carbon parameters should be defined the
+same way as `let` above. The C++ functions:
+
+```
+auto LogSize(const Container &large_data) {
+  Logger.Print(large_data.Size());
+}
+
+auto Sum(const int x, const int y) -> int {
+  return x + y;
+}
+```
+
+would become Carbon code such as:
+
+```
+fn LogSize(large_data: Container) {
```

Comment:
This seems to take `Container` by value; how does this not require the caller to copy a `Container` (or, alternatively, how does a function take a `large_data`'s ownership)

Reply:
It seems specifically here that Carbon is avoiding ownership types a la Rust/C++, which I don't personally love... How do you plan on doing ownership?